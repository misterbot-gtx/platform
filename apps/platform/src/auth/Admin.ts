import { OrganizationRole } from '../organizations/Organization'
import Model, { ModelParams } from '../core/Model'

export default class Admin extends Model {
    organization_id!: number
    email!: string
    first_name?: string
    last_name?: string
    image_url?: string
    role!: OrganizationRole
}

export type AdminParams = Omit<Admin, ModelParams> & { domain?: string }

export type AuthAdminParams = Omit<AdminParams, 'organization_id' | 'role'> & { domain?: string }
