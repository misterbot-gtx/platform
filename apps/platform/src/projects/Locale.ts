import Model, { ModelParams } from '../core/Model'

export default class Locale extends Model {
    project_id!: number
    key!: string
    label!: string
}

export type LocaleParams = Omit<Locale, ModelParams | 'project_id'>
