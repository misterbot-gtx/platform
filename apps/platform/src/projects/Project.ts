import Model, { ModelParams } from '../core/Model'

export default class Project extends Model {
    organization_id!: number
    name!: string
    description?: string
    deleted_at?: Date
    locale!: string
    timezone!: string
    text_opt_out_message?: string
    text_help_message?: string
    link_wrap_email?: boolean
    link_wrap_push?: boolean
}

export type ProjectParams = Omit<Project, ModelParams | 'deleted_at' | 'organization_id'>

export const projectRoles = [
    'support',
    'editor',
    'admin',
] as const

export type ProjectRole = (typeof projectRoles)[number]
