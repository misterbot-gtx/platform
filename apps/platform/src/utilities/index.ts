import { JSONSchemaType } from 'ajv'
import { validate } from '../core/validate'
import crypto from 'crypto'
import Hashids from 'hashids'
import { differenceInSeconds } from 'date-fns'
import { utcToZonedTime } from 'date-fns-tz'
import { Database } from '../config/database'

export const pluralize = (noun: string, count = 2, suffix = 's') => `${noun}${count !== 1 ? suffix : ''}`

export const random = <T>(array: T[]): T => array[Math.floor(Math.random() * array.length)]

export const shuffle = <T>(array: T[]): T[] => {
    let currentIndex = array.length
    let randomIndex: number
    while (currentIndex > 0) {
        randomIndex = Math.floor(Math.random() * currentIndex)
        currentIndex--;
        [array[currentIndex], array[randomIndex]] = [array[randomIndex], array[currentIndex]]
    }
    return array
}

export const pick = <T extends object, K extends keyof T> (obj: T, keys: K[]): Pick<T, K> => {
    const ret: any = {}
    keys.forEach(key => {
        ret[key] = obj[key]
    })
    return ret
}

export const randomInt = (min = 0, max = 100): number => {
    min = Math.ceil(min)
    max = Math.floor(max)
    return Math.floor(Math.random() * (max - min + 1)) + min
}

export const prune = (obj: Record<string, any>): Record<string, any> => {
    return Object.fromEntries(
        Object.entries(obj)
            .filter(([_, v]) => v != null && v !== '')
            .map(([k, v]) => [k, v === Object(v) ? prune(v) : v]),
    )
}

export const isValidUrl = (url: string) => {
    try {
        return Boolean(new URL(url))
    } catch {
        return false
    }
}

export const snakeCase = (str: string): string => str.match(/[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g)
    ?.map(x => x.toLowerCase())
    .join('_') ?? ''

export const uuid = (): string => {
    return crypto.randomUUID()
}

export const secondsAgo = (timestamp: Date | number) => {
    return differenceInSeconds(Date.now(), timestamp)
}

const hashCharacters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890'
const hashMinimumLength = 10
export const encodeHashid = function(value: number): string {
    const hashids = new Hashids(
        process.env.APP_SECRET,
        hashMinimumLength,
        hashCharacters,
    )
    return hashids.encode(value)
}

export const decodeHashid = function(value?: string | null): number | undefined {
    if (!value) return
    const hashids = new Hashids(
        process.env.APP_SECRET,
        hashMinimumLength,
        hashCharacters,
    )
    return hashids.decode(value)?.[0] as number | undefined
}

export const combineURLs = (parts: string[], sep = '/'): string => {
    return parts
        .map(part => {
            const part2 = part.endsWith(sep) ? part.substring(0, part.length - 1) : part
            return part2.startsWith(sep) ? part2.substr(1) : part2
        })
        .join(sep)
}

export const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))

export const removeKey = <T, O extends keyof T>(propKey: O, { [propKey]: propValue, ...rest }: T): Omit<T, O> => rest

export const batch = <T>(arr: T[], size: number) => {
    const result: T[][] = []
    for (let i = 0, len = arr.length; i < len; i += size) {
        result.push(arr.slice(i, i + size))
    }
    return result
}

export function findLast<T>(arr: T[], predicate: (item: T, index: number, arr: T[]) => boolean) {
    for (let i = arr.length - 1; i >= 0; i--) {
        const item = arr[i]
        if (predicate(item, i, arr)) {
            return item
        }
    }
    return undefined
}

export const groupBy = <T, K>(arr: T[], getKey: (o: T) => K) => {
    const m = new Map<K, T[]>()
    for (const item of arr) {
        const key = getKey(item)
        let list = m.get(key)
        if (!list) m.set(key, list = [])
        list.push(item)
    }
    return m
}

export const parseLocale = (locale: string): string | undefined => {
    return locale.slice(-1) === '-'
        ? locale.slice(0, -1)
        : locale
}

export const partialMatchLocale = (locale1?: string, locale2?: string) => {
    const locale1Root = locale1?.split('-')[0]
    const locale2Root = locale2?.split('-')[0]
    return locale1 === locale2 || locale1Root === locale2Root
}

export const isValidIANATimezone = (timezone?: string) => {
    try {
        if (!timezone || typeof timezone !== 'string') {
            return false
        }

        Intl.DateTimeFormat(undefined, { timeZone: timezone })
        return true
    } catch (e) {
        return false
    }
}

export const crossTimezoneCopy = (
    date: Date,
    fromTimezone: string,
    toTimezone: string,
) => {
    const baseDate = utcToZonedTime(date, fromTimezone)

    const utcDate = new Date(baseDate.toLocaleString('en-US', { timeZone: 'UTC' }))
    const tzDate = new Date(baseDate.toLocaleString('en-US', { timeZone: toTimezone }))
    const offset = utcDate.getTime() - tzDate.getTime()

    baseDate.setTime(baseDate.getTime() + offset)
    return baseDate
}

export function extractQueryParams<T extends Record<string, any>>(search: URLSearchParams | Record<string, undefined | string | string[]>, schema: JSONSchemaType<T>) {
    return validate(schema, Object.entries<JSONSchemaType<any>>(schema.properties).reduce((a, [name, def]) => {
        let values!: string[]

        if (search instanceof URLSearchParams) {
            values = search.getAll(name)
        } else {
            const v = search[name]
            values = Array.isArray(v) ? v : v ? [v] : []
        }
        if (def.type !== 'array' && values.length > 1) {
            values = values.slice(0, 1)
        }
        if (values.length) {
            const transformed = values.map(v => {
                const type = def.type === 'array' ? def.items.type : def.type
                if (type === 'boolean') {
                    return Boolean(v && 'ty1'.includes(v.charAt(0).toLowerCase()))
                } else if (type === 'integer') {
                    const i = parseInt(v, 10)
                    return isNaN(i) ? undefined : i
                } else if (type === 'number') {
                    const f = parseFloat(v)
                    return isNaN(f) ? undefined : f
                } else if (type === 'array' || type === 'object') {
                    return JSON.parse(v)
                } else {
                    return v
                }
            })
            if (def.type === 'array') {
                a[name] = transformed
            } else {
                a[name] = transformed.at(0)
            }
        }
        if (a[name] === undefined) {
            a[name] = def.default
        }
        return a
    }, {} as Record<string, any>))
}

export function firstQueryParam(value: string | string[] | undefined) {
    return Array.isArray(value) ? value[0] : value
}

export function capitalizeFirstLetter(string: string) {
    return string.charAt(0).toUpperCase() + string.slice(1)
}

export function shallowEqual(object1: any, object2: any) {
    if (object1 === object2) return true
    if (!object1 || !object2) return false

    const keys1 = Object.keys(object1)
    const keys2 = Object.keys(object2)

    if (keys1.length !== keys2.length) return false
    return keys1.every(key => object1[key] === object2[key])
}

type ChunkCallback<T> = (chunk: T[]) => Promise<void>

export const chunk = async <T>(
    query: Database.QueryBuilder,
    size = 100,
    callback: ChunkCallback<T>,
    modifier: (result: any) => T = (result) => result,
) => {
    const chunker = new Chunker(callback, size)
    const handler = async (result: any, retries = 3) => {
        try {
            await chunker.add(modifier(result))
        } catch (error: any) {

            // In the case of deadlocks, retry the operation
            if (['ER_LOCK_WAIT_TIMEOUT', 'ER_LOCK_DEADLOCK'].includes(error.code) && retries > 0) {
                setTimeout(() => handler(result, retries - 1), 250)
            } else {
                throw error
            }
        }
    }
    await query.stream(async function(stream) {
        for await (const result of stream) {
            await handler(result)
        }
    })
    await chunker.flush()
}

export class Chunker<T> {

    #items: T[] = []

    constructor(
        private callback: (batch: T[]) => Promise<void>,
        private size: number,
    ) {}

    public async add(...items: T[]) {
        for (const item of items) {
            this.#items.push(item)
            if (this.#items.length >= this.size) {
                await this.flush()
            }
        }
    }

    public async flush() {
        if (this.#items.length) {
            await this.callback(this.#items)
            this.#items = []
        }
    }
}

export function visit<T>(item: T, children: (item: T) => undefined | T[], callback: (item: T) => void) {
    callback(item)
    const items = children(item)
    if (items?.length) {
        for (const item of items) {
            visit(item, children, callback)
        }
    }
}
