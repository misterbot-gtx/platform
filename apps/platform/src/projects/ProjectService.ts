import { ProjectState } from '../auth/AuthMiddleware'
import { Next, ParameterizedContext } from 'koa'
import { RequestError } from '../core/errors'
import { PageParams } from '../core/searchParams'
import { createSubscription } from '../subscriptions/SubscriptionService'
import { uuid } from '../utilities'
import Project, { ProjectParams, ProjectRole, projectRoles } from './Project'
import { ProjectAdmin } from './ProjectAdmins'
import { ProjectApiKey, ProjectApiKeyParams } from './ProjectApiKey'
import Admin from '../auth/Admin'
import { getAdmin } from '../auth/AdminRepository'
import Locale, { LocaleParams } from './Locale'

export const adminProjectIds = async (adminId: number) => {
    const records = await ProjectAdmin.all(qb => qb.where('admin_id', adminId))
    return records.map(item => item.project_id)
}

export const pagedProjects = async (params: PageParams, adminId: number, organizationId: number) => {
    const admin = await getAdmin(adminId, organizationId)
    const projectIds = await adminProjectIds(adminId)
    return await Project.search({ ...params, fields: ['name'] }, qb =>
        qb.where(qb =>
            qb.where('organization_id', admin!.organization_id)
                .orWhereIn('projects.id', projectIds),
        ),
    )
}

export const allProjects = async (adminId: number, organizationId: number) => {
    const admin = await getAdmin(adminId, organizationId)
    if (!admin) return []
    const projectIds = await adminProjectIds(adminId)
    return await Project.all(qb => {
        qb.whereIn('projects.id', projectIds)
        if (admin.role !== 'member') {
            qb.orWhere('organization_id', admin.organization_id)
        }
        return qb
    })
}

export const getProject = async (id: number, adminId?: number) => {
    return Project.first(
        qb => {
            qb.where('projects.id', id)
                .select('projects.*')
            if (adminId != null) {
                qb.leftJoin('project_admins', 'project_admins.project_id', 'projects.id')
                    .where('admin_id', adminId)
                    .select('role')
            }
            return qb
        })
}

export const createProject = async (admin: Admin, params: ProjectParams): Promise<Project> => {
    const projectId = await Project.insert({
        ...params,
        organization_id: admin.organization_id,
    })

    // Add the user creating the project to it
    await ProjectAdmin.insert({
        project_id: projectId,
        admin_id: admin.id,
        role: 'admin',
    })

    // Create initial locale for project
    const languages = new Intl.DisplayNames([params.locale], { type: 'language' })
    await Locale.insert({
        project_id: projectId,
        key: params.locale,
        label: languages.of(params.locale),
    })

    // Create a single subscription for each type
    await createSubscription(projectId, { name: 'Default Email', channel: 'email' })
    await createSubscription(projectId, { name: 'Default SMS', channel: 'text' })
    await createSubscription(projectId, { name: 'Default Push', channel: 'push' })
    await createSubscription(projectId, { name: 'Default Webhook', channel: 'webhook' })

    const project = await getProject(projectId, admin.id)
    return project!
}

export const updateProject = async (id: number, adminId: number, params: Partial<ProjectParams>) => {
    await Project.update(qb => qb.where('id', id), params)
    return await getProject(id, adminId)
}

export const pagedApiKeys = async (params: PageParams, projectId: number) => {
    return await ProjectApiKey.search(
        { ...params, fields: ['name', 'description'] },
        qb => qb.where('project_id', projectId).whereNull('deleted_at'),
    )
}

export const getProjectApiKey = async (key: string) => {
    return ProjectApiKey.first(qb => qb.where('value', key).whereNull('deleted_at'))
}

export const createProjectApiKey = async (projectId: number, params: ProjectApiKeyParams) => {
    return await ProjectApiKey.insertAndFetch({
        ...params,
        value: generateApiKey(params.scope),
        project_id: projectId,
    })
}

export const updateProjectApiKey = async (id: number, params: ProjectApiKeyParams) => {
    return await ProjectApiKey.updateAndFetch(id, params)
}

export const revokeProjectApiKey = async (id: number) => {
    return await ProjectApiKey.archive(id)
}

export const generateApiKey = (scope: 'public' | 'secret') => {
    const key = uuid().replace('-', '')
    const prefix = scope === 'public' ? 'pk' : 'sk'
    return `${prefix}_${key}`
}

export const requireProjectRole = (ctx: ParameterizedContext<ProjectState>, minRole: ProjectRole) => {
    if (projectRoles.indexOf(minRole) > projectRoles.indexOf(ctx.state.projectRole)) {
        throw new RequestError(`Minimum project role ${minRole} is required`, 403)
    }
}

export const projectRoleMiddleware = (minRole: ProjectRole) => async (ctx: ParameterizedContext<ProjectState>, next: Next) => {
    requireProjectRole(ctx, minRole)
    return next()
}

export const pagedLocales = async (params: PageParams, projectId: number) => {
    return await Locale.search(
        { ...params, fields: ['name'] },
        qb => qb.where('project_id', projectId),
    )
}

export const createLocale = async (projectId: number, params: LocaleParams) => {
    return await Locale.insertAndFetch({
        ...params,
        project_id: projectId,
    })
}

export const deleteLocale = async (projectId: number, id: number) => {
    return await Locale.deleteById(id, qb => qb.where('project_id', projectId))
}
