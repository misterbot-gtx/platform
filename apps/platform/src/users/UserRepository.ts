import { RuleTree } from '../rules/Rule'
import { ClientAliasParams, ClientIdentity } from '../client/Client'
import { PageParams } from '../core/searchParams'
import { RetryError } from '../queue/Job'
import { subscribeAll } from '../subscriptions/SubscriptionService'
import { Device, DeviceParams, User, UserInternalParams } from '../users/User'
import { uuid } from '../utilities'
import { getRuleEventNames } from '../rules/RuleHelpers'
import { UserEvent } from './UserEvent'
import { Context } from 'koa'
import { EventPostJob } from '../jobs'

export const getUser = async (id: number, projectId?: number): Promise<User | undefined> => {
    return await User.find(id, qb => {
        if (projectId) {
            qb.where('project_id', projectId)
        }
        return qb
    })
}

export const getUserFromContext = async (ctx: Context): Promise<User | undefined> => {
    return ctx.state.scope === 'secret'
        ? await getUserFromClientId(ctx.state.project.id, { external_id: ctx.params.userId })
        : await getUser(parseInt(ctx.params.userId), ctx.state.project.id)
}

export const getUsersFromIdentity = async (projectId: number, identity: ClientIdentity) => {
    const externalId = `${identity.external_id}`
    const anonymousId = `${identity.anonymous_id}`

    const users = await User.all(
        qb => qb
            .where(sqb => {
                if (identity.external_id) {
                    sqb.where('external_id', externalId)
                }
                if (identity.anonymous_id) {
                    sqb.orWhere('anonymous_id', anonymousId)
                }
            })
            .where('project_id', projectId)
            .limit(2),
    )

    // Map each ID to a key so they are both available
    return {
        anonymous: users.find(user => user.anonymous_id === anonymousId),
        external: users.find(user => user.external_id === externalId),
    }
}

export const getUserFromClientId = async (projectId: number, identity: ClientIdentity): Promise<User | undefined> => {
    const users = await getUsersFromIdentity(projectId, identity)

    // There are circumstances in which both the external ID and
    // the anonymous ID match but match different records in
    // those cases, default to the one matching the external ID
    return users.external ?? users.anonymous
}

export const getUserFromPhone = async (projectId: number, phone: string): Promise<User | undefined> => {
    return await User.first(
        qb => qb.where('phone', phone)
            .where('project_id', projectId),
    )
}

export const getUserFromEmail = async (projectId: number, email: string): Promise<User | undefined> => {
    return await User.first(
        qb => qb.where('email', email)
            .where('project_id', projectId),
    )
}

export const pagedUsers = async (params: PageParams, projectId: number) => {
    return await User.search(
        {
            ...params,
            fields: ['external_id', 'email', 'phone'],
            mode: 'exact',
        },
        b => b.where('project_id', projectId),
    )
}

export const aliasUser = async (projectId: number, {
    external_id,
    anonymous_id,
    previous_id,
}: ClientAliasParams): Promise<User | undefined> => {

    // Previous is the one merging into userId
    const previous = await getUserFromClientId(projectId, {
        external_id: previous_id,
        anonymous_id,
    } as ClientIdentity)
    if (!previous) return

    // Look up if there is a separate profile with the new ID
    // If there is one, update that clients anonymous ID
    const current = await getUserFromClientId(projectId, {
        external_id,
    } as ClientIdentity)
    if (current) return

    return await User.updateAndFetch(previous.id, { external_id })
}

export const createUser = async (projectId: number, { external_id, anonymous_id, data, created_at, ...fields }: UserInternalParams) => {
    const user = await User.insertAndFetch({
        project_id: projectId,
        anonymous_id: anonymous_id ?? uuid(),
        external_id,
        data: data ?? {},
        created_at: created_at ? new Date(created_at) : new Date(),
        ...fields,
    })

    // Subscribe the user to all channels the user has available
    await subscribeAll(user)

    // Create an event for the user creation
    await EventPostJob.from({
        project_id: projectId,
        user_id: user.id,
        event: {
            name: 'user_created',
            external_id: user.external_id,
            anonymous_id,
            data: { ...data, ...fields, external_id, anonymous_id },
        },
    }).queue()

    return user
}

export const saveDevice = async (projectId: number, { external_id, anonymous_id, ...params }: DeviceParams): Promise<Device | undefined> => {

    const user = await getUserFromClientId(projectId, { external_id, anonymous_id } as ClientIdentity)
    if (!user) throw new RetryError()

    let isFirstDevice = false
    if (!user.devices) {
        user.devices = []
        isFirstDevice = true
    }
    let device = user.devices?.find(device => {
        return device.device_id === params.device_id
            || (device.token === params.token && device.token != null)
    })
    if (device) {
        Object.assign(device, params)
    } else {
        device = {
            ...params,
            device_id: params.device_id,
        }
        user.devices.push(device)
    }
    await User.updateAndFetch(user.id, { devices: user.devices })

    if (isFirstDevice) {
        await subscribeAll(user, ['push'])
    }
    return device
}

export const disableNotifications = async (userId: number, tokens: string[]): Promise<boolean> => {
    const user = await User.find(userId)
    if (!user) return false
    const device = user.devices?.find(device => device.token && tokens.includes(device.token))
    if (device) device.token = undefined
    await User.update(qb => qb.where('id', userId), {
        devices: user.devices,
    })
    return true
}

export const getUserEventsForRules = async (
    userIds: number[],
    rules: RuleTree[],
    since?: Date | null,
) => {
    if (!userIds.length || !rules.length) return []
    const names = rules.reduce<string[]>((a, rule) => {
        if (rule) {
            a.push(...getRuleEventNames(rule))
        }
        return a
    }, []).filter((o, i, a) => a.indexOf(o) === i)
    if (!names.length) return []
    return UserEvent.all(qb => {
        qb.whereIn('user_id', userIds)
            .whereIn('name', names)
            .orderBy('id', 'asc')
        if (since) qb.where('created_at', '>=', since)
        return qb
    })
}

export const isUserDirty = (params: UserInternalParams) => {
    const hasData = !!params.data && Object.keys(params.data).length > 0
    const hasReserved = !!params.email || !!params.phone || !!params.timezone || !!params.locale || !!params.created_at

    return hasData || hasReserved
}
