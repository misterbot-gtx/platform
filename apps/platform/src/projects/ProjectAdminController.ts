import Router from '@koa/router'
import { extractQueryParams } from '../utilities'
import { searchParamsSchema } from '../core/searchParams'
import { ProjectState } from '../auth/AuthMiddleware'
import { addAdminToProject, getProjectAdmin, pagedProjectAdmins, removeAdminFromProject } from './ProjectAdminRepository'
import { JSONSchemaType } from 'ajv'
import Admin from '../auth/Admin'
import { validate } from '../core/validate'
import { projectRoleMiddleware } from './ProjectService'
import { ProjectAdminParams } from './ProjectAdmins'
import { projectRoles } from './Project'
import { RequestError } from '../core/errors'
import { createOrUpdateAdmin } from '../auth/AdminRepository'

const router = new Router<
    ProjectState & { admin?: Admin }
>({
    prefix: '/admins',
})

router.use(projectRoleMiddleware('admin'))

router.get('/', async ctx => {
    const params = extractQueryParams(ctx.query, searchParamsSchema)
    ctx.body = await pagedProjectAdmins(params, ctx.state.project.id)
})

const projectCreateAdminParamsSchema: JSONSchemaType<ProjectAdminParams & { email: string }> = {
    $id: 'projectCreateAdminParams',
    type: 'object',
    required: ['role', 'email'],
    properties: {
        email: {
            type: 'string',
            format: 'email',
        },
        role: {
            type: 'string',
            enum: projectRoles,
        },
    },
}

router.post('/', async ctx => {
    const organizationId = ctx.state.project.organization_id
    const { role, email } = validate(projectCreateAdminParamsSchema, ctx.request.body)
    const admin = await createOrUpdateAdmin({
        organization_id: organizationId,
        email,
        role: 'member',
    })
    if (ctx.state.admin!.id === admin.id) throw new RequestError('You cannot add yourself to a project')
    await addAdminToProject(ctx.state.project.id, admin.id, role)
    ctx.body = await getProjectAdmin(ctx.state.project.id, admin.id)
})

const projectAdminParamsSchema: JSONSchemaType<ProjectAdminParams> = {
    $id: 'projectAdminParams',
    type: 'object',
    required: ['role'],
    properties: {
        role: {
            type: 'string',
            enum: projectRoles,
        },
    },
}

router.put('/:adminId', async ctx => {
    const admin = await Admin.find(ctx.params.adminId)
    if (!admin) throw new RequestError('Invalid admin ID', 404)
    const { role } = validate(projectAdminParamsSchema, ctx.request.body)
    if (ctx.state.admin!.id === admin.id) throw new RequestError('You cannot add yourself to a project')
    await addAdminToProject(ctx.state.project.id, admin.id, role)
    ctx.body = await getProjectAdmin(ctx.state.project.id, admin.id)
})

router.get('/:adminId', async ctx => {
    const projectAdmin = await getProjectAdmin(ctx.state.project.id, parseInt(ctx.params.adminId, 10))
    if (!projectAdmin) return ctx.throw(404)
    ctx.body = projectAdmin
})

router.delete('/:adminId', async ctx => {
    await removeAdminFromProject(ctx.state.project.id, parseInt(ctx.params.adminId, 10))
    ctx.body = true
})

export default router
