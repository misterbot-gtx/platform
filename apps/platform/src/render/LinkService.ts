import { URL } from 'node:url'
import App from '../app'
import Campaign from '../campaigns/Campaign'
import CampaignInteractJob from '../campaigns/CampaignInteractJob'
import { getCampaign } from '../campaigns/CampaignService'
import EventPostJob from '../client/EventPostJob'
import { User } from '../users/User'
import { getUser } from '../users/UserRepository'
import { combineURLs, decodeHashid, encodeHashid } from '../utilities'

export interface TrackedLinkParams {
    userId: number
    campaignId: number
    referenceId?: string
}

interface TrackedLinkParts extends TrackedLinkParams {
    path: string
    redirect?: string
}

export const paramsToEncodedLink = (params: TrackedLinkParts): string => {
    const hashUserId = encodeHashid(params.userId)
    const hashCampaignId = encodeHashid(params.campaignId)

    const baseUrl = combineURLs([App.main.env.baseUrl, params.path])
    const url = new URL(baseUrl)
    url.searchParams.set('u', hashUserId)
    url.searchParams.set('c', hashCampaignId)
    if (params.referenceId) {
        url.searchParams.set('s', params.referenceId)
    }
    if (params.redirect) {
        url.searchParams.set('r', encodeURIComponent(params.redirect))
    }
    return url.href
}

interface TrackedLinkExport {
    user?: User
    campaign?: Campaign
    referenceId?: string
    redirect: string
}

export const encodedLinkToParts = async (link: string | URL): Promise<TrackedLinkExport> => {
    const url = link instanceof URL ? link : new URL(link)
    const userId = decodeHashid(url.searchParams.get('u'))
    const campaignId = decodeHashid(url.searchParams.get('c'))
    const referenceId = url.searchParams.get('s') ?? undefined
    const redirect = decodeURIComponent(url.searchParams.get('r') ?? '')

    const parts: TrackedLinkExport = { redirect, referenceId }

    if (userId) {
        parts.user = await getUser(userId)

        if (parts.user && campaignId) {
            parts.campaign = await getCampaign(campaignId, parts.user.project_id)
        }
    }

    return parts
}

export const clickWrapHtml = (html: string, params: TrackedLinkParams) => {
    const regex = /a.*\s*href\s*=\s*(['"])(https?:\/\/.+?)\1/gi
    let link

    while ((link = regex.exec(html)) !== null) {
        const redirect = link[2]

        html = html.replace(
            redirect,
            paramsToEncodedLink({ ...params, redirect, path: 'c' }),
        )
    }

    return html
}

export const openWrapHtml = (html: string, params: TrackedLinkParams) => {
    const link = paramsToEncodedLink({ ...params, path: 'o' })
    const imageHtml = `<img border="0" width="1" height="1" alt="" src="${link}" />`
    return injectInBody(html, imageHtml, 'end')
}

export const preheaderWrapHtml = (html: string, preheader: string) => {
    const preheaderHtml = `<span style="color:transparent;visibility:hidden;display:none;opacity:0;height:0;width:0;font-size:0">${preheader}</span>`
    return injectInBody(html, preheaderHtml, 'start')
}

export const injectInBody = (html: string, injection: string, placement: 'start' | 'end') => {
    if (placement === 'end') {
        const bodyTag = '</body'
        html = html.includes(bodyTag)
            ? html = html.replace(bodyTag, (injection + bodyTag))
            : html += injection
    } else {
        const regex = /<body(?:.|\n)*?>/
        const match = html.match(regex)
        if (match) {
            const offset = match.index! + match[0].length
            html = html.substring(0, offset) + injection + html.substring(offset)
        } else {
            html = injection + html
        }
    }
    return html
}

export type TrackMessageType = 'opened' | 'clicked' | 'bounced' | 'complained' | 'failed'
export const trackMessageEvent = async (
    parts: Partial<TrackedLinkExport>,
    type: TrackMessageType,
    action?: 'unsubscribe',
    context?: any,
) => {
    const { user, campaign, referenceId } = parts
    if (!user || !campaign) return

    const eventJob = EventPostJob.from({
        project_id: user.project_id,
        event: {
            external_id: user.external_id,
            name: campaign.eventName(type),
            data: {
                campaign_id: campaign.id,
                campaign_name: campaign.name,
                campaign_type: campaign.type,
                channel: campaign.channel,
                subscription_id: campaign.subscription_id,
                url: parts.redirect,
                context,
            },
        },
        forward: true,
    })

    const campaignJob = CampaignInteractJob.from({
        campaign_id: campaign.id,
        user_id: user.id,
        reference_id: referenceId ?? '0',
        subscription_id: campaign.subscription_id,
        type,
        action,
    })

    await App.main.queue.enqueueBatch([
        eventJob,
        campaignJob,
    ])
}
