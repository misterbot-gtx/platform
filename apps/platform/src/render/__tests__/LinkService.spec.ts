import Admin from '../../auth/Admin'
import { createProject } from '../../projects/ProjectService'
import { createUser } from '../../users/UserRepository'
import { uuid } from '../../utilities'
import { clickWrapHtml, encodedLinkToParts, openWrapHtml, paramsToEncodedLink, preheaderWrapHtml } from '../LinkService'

describe('LinkService', () => {
    describe('encodedLinkToParts', () => {
        test('a properly encoded link decodes to parts', async () => {
            const admin = await Admin.insertAndFetch({
                first_name: uuid(),
                last_name: uuid(),
                email: `${uuid()}@test.com`,
            })
            const project = await createProject(admin, {
                name: uuid(),
                timezone: 'utc',
                locale: 'en',
            })
            const user = await createUser(project.id, {
                anonymous_id: uuid(),
                external_id: uuid(),
            })

            const redirect = 'http://test.com'
            const link = paramsToEncodedLink({
                userId: user.id,
                campaignId: 1,
                path: 'c',
                redirect,
            })

            const parts = await encodedLinkToParts(link)
            expect(parts.user?.id).toEqual(user.id)
            expect(parts.redirect).toEqual(redirect)
        })
    })

    describe('clickWrapHtml', () => {
        test('links are wrapped', async () => {
            const html = 'This is some html <a href="https://test.com">Test Link</a>'
            const output = clickWrapHtml(html, { userId: 1, campaignId: 2 })
            expect(output).toMatchSnapshot()
        })
    })

    describe('openWrapHtml', () => {
        test('open tracking image is added to end of body', async () => {
            const html = '<html><body>This is some html</body></html>'
            const output = openWrapHtml(html, { userId: 3, campaignId: 4 })
            expect(output).toMatchSnapshot()
        })
    })

    describe('preheaderWrapHtml', () => {
        test('simple html injects preheader', async () => {
            const preheader = 'This is some preheader'
            const html = '<html><body>This is some html</body></html>'
            const output = preheaderWrapHtml(html, preheader)
            expect(output).toMatchSnapshot()
        })

        test('complex html injects preheader', async () => {
            const preheader = 'This is some preheader'
            const html = `<html>
                <body
                    style="color:#000"
                    class="body-class">
                    This is some html
                </body>
            </html>`
            const output = preheaderWrapHtml(html, preheader)
            expect(output).toMatchSnapshot()
        })
    })
})
