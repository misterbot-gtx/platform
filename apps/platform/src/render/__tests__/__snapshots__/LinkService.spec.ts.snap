// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkService clickWrapHtml links are wrapped 1`] = `"This is some html <a href=\\"https://parcelvoy.com/c?u=M8LRMWZ645&c=1MLXN7R524&r=https%253A%252F%252Ftest.com\\">Test Link</a>"`;

exports[`LinkService openWrapHtml open tracking image is added to end of body 1`] = `"<html><body>This is some html<img border=\\"0\\" width=\\"1\\" height=\\"1\\" alt=\\"\\" src=\\"https://parcelvoy.com/o?u=P9KR98X8L4&c=Y3QRV4XL74\\" /></body></html>"`;

exports[`LinkService preheaderWrapHtml complex html injects preheader 1`] = `
"<html>
                <body
                    style=\\"color:#000\\"
                    class=\\"body-class\\"><span style=\\"color:transparent;visibility:hidden;display:none;opacity:0;height:0;width:0;font-size:0\\">This is some preheader</span>
                    This is some html
                </body>
            </html>"
`;

exports[`LinkService preheaderWrapHtml simple html injects preheader 1`] = `"<html><body><span style=\\"color:transparent;visibility:hidden;display:none;opacity:0;height:0;width:0;font-size:0\\">This is some preheader</span>This is some html</body></html>"`;
